File: frontend/src/dsp/pf_rul.ts
// Particle-filter RUL prognostics, proper Bayesian state estimation over the exponential
// degradation model HI(t) = a·exp(b·t). Unlike the classical OLS fit, the PF maintains a full
// posterior distribution over (ln a, b, σ_obs) that is updated sequentially as each new HI
// observation arrives. This buys:
//   - calibrated uncertainty that grows/shrinks with data density,
//   - robustness to noise (the filter tracks the state, not the raw HI),
//   - a genuine posterior RUL distribution from the particle ensemble.
//
// Algorithm: regularised auxiliary SIR with kernel-density jitter (Musso, Oudjane & Le Gland 2001)
// and a wide uninformed prior (not seeded from OLS, the filter must earn its estimate from data).
//
// References:
//   Arulampalam, Maskell, Gordon & Clapp (2002), IEEE TSP 50(2):174–188
//   Musso, Oudjane & Le Gland (2001), "Improving regularised particle filters", pp. 247–271
//   An, Kim & Choi (2013), RESS 120:50–62
//   Orchard & Vachtsevanos (2009), T I Measurement & Control 31(3-4):221–246
import { type HIPoint } from './health';

// ── types ───────────────────────────────────────────────────────────────────
export interface Particle { lnA: number; b: number; sigmaObs: number; w: number }
export interface PfRulResult {
  onset: number | null;
  rulMedian: number | null;
  rulP10: number | null;
  rulP90: number | null;
  failTimeMedian: number | null;
  particles: Particle[];
  rulEnsemble: number[];
  converged: boolean;
}

const N = 500;                 // particles
const MIN_POST_ONSET = 6;     // minimum post-onset observations needed
const ESS_THRESHOLD = 0.60;   // resample when ESS/N drops below this

// ── kernel-density regularisation (Musso et al. 2001) ──────────────────────
function kernelRegularise(p: Particle[], rng: Rng): void {
  const n = p.length;
  // Silverman bandwidth for each dimension (using weighted samples)
  const wSum = p.reduce((a, v) => a + v.w, 0) || 1;
  const mean = [0, 0, 0];
  for (const x of p) { mean[0] += x.lnA * x.w / wSum; mean[1] += x.b * x.w / wSum; mean[2] += x.sigmaObs * x.w / wSum; }
  let varL = 0, varB = 0, varS = 0;
  for (const x of p) {
    varL += x.w / wSum * (x.lnA - mean[0]) ** 2;
    varB += x.w / wSum * (x.b - mean[1]) ** 2;
    varS += x.w / wSum * (x.sigmaObs - mean[2]) ** 2;
  }
  // Silverman: h = 1.06 * σ * N^(-1/5), then shrink by 0.5 (regularisation, not full kernel)
  const hL = Math.sqrt(Math.max(1e-9, varL)) * 1.06 * Math.pow(n, -0.2) * 0.5;
  const hB = Math.sqrt(Math.max(1e-16, varB)) * 1.06 * Math.pow(n, -0.2) * 0.5;
  const hS = Math.sqrt(Math.max(1e-12, varS)) * 1.06 * Math.pow(n, -0.2) * 0.5;
  for (const x of p) {
    x.lnA += randn(rng) * hL;
    x.b = Math.max(1e-12, x.b + randn(rng) * hB);
    x.sigmaObs = Math.max(1e-6, x.sigmaObs + randn(rng) * hS);
  }
}

// ── systematic resampling ───────────────────────────────────────────────────
function resample(p: Particle[], rng: Rng): void {
  const n = p.length;
  const ws = p.map(x => x.w);
  const sum = ws.reduce((a, v) => a + v, 0) || 1;
  for (let i = 0; i < n; i++) ws[i] /= sum;
  const invN = 1 / n;
  const u0 = rng() * invN;
  const out: Particle[] = new Array(n);
  let c = ws[0], j = 0;
  for (let i = 0; i < n; i++) {
    const u = u0 + i * invN;
    while (u > c && j < n - 1) { j++; c += ws[j]; }
    const src = p[j];
    out[i] = { lnA: src.lnA, b: src.b, sigmaObs: src.sigmaObs, w: invN };
  }
  p.length = 0; p.push(...out);
}

// ── the random source ───────────────────────────────────────────────────────
//
// This filter used to draw from Math.random directly, in three places: the prior particles, the
// kernel-regularisation jitter and the systematic-resampling offset. Nothing could pin it, so every
// assertion about the posterior was a coin flip. MEASURED on frontend/test/pf_test.ts before this
// change: 4 failures in 20 runs (posterior sd 0.81, 0.83, 0.83 and 1.27 against a `< 0.8` bound).
//
// `Rng` is just `() => number` in [0,1). The default is Math.random, so the browser behaves exactly as
// it did and nothing already baked changes; passing a seed makes a run reproducible. The Python twin
// in data-pipeline/pipeline/model/pf_rul.py takes the same optional seed, which is what keeps the
// cross-validation between the two lanes meaningful.
type Rng = () => number;

/** mulberry32: small, fast, and good enough for a particle filter's proposal noise. */
function mulberry32(seed: number): Rng {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let x = Math.imul(a ^ (a >>> 15), 1 | a);
    x = (x + Math.imul(x ^ (x >>> 7), 61 | x)) ^ x;
    return ((x ^ (x >>> 14)) >>> 0) / 4294967296;
  };
}

// ── Box-Muller ──────────────────────────────────────────────────────────────
function randn(rng: Rng): number {
  let u = 0, v = 0;
  while (u === 0) u = rng();
  while (v === 0) v = rng();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

// ── log-likelihood ──────────────────────────────────────────────────────────
function loglik(obs: number, lnA: number, b: number, sigma: number, t: number): number {
  const pred = lnA + b * t;
  const err = Math.log(Math.max(1e-9, obs)) - pred;
  return -0.5 * (err / sigma) ** 2 - Math.log(sigma);
}

// ── onset detection (statistical, same as classical, robust to noise) ─────
function detectOnset(points: HIPoint[]): number | null {
  const n = points.length;
  if (n < 10) return null;
  const baseN = Math.max(5, Math.floor(n * 0.25));
  const base = points.slice(0, baseN);
  const mu0 = base.reduce((a, p) => a + p.hi, 0) / base.length;
  const sd0 = Math.sqrt(base.reduce((a, p) => a + (p.hi - mu0) ** 2, 0) / base.length) || 1e-9;
  // Adaptive threshold: 3.5σ instead of 4σ to handle real noisy data
  const thr = mu0 + 3.5 * sd0;
  for (let i = 1; i < n - 2; i++) {
    if (points[i].hi > thr && points[i + 1].hi > thr && points[i + 2].hi > thr) {
      return i;
    }
  }
  return null;
}

// ── main ────────────────────────────────────────────────────────────────────
export function particleFilterRUL(points: HIPoint[], threshold: number, seed?: number): PfRulResult {
  // undefined seed keeps the previous behaviour exactly; a seed makes the posterior reproducible.
  const rng: Rng = seed === undefined ? Math.random : mulberry32(seed);
  const n = points.length;
  const nope: PfRulResult = {
    onset: null, rulMedian: null, rulP10: null, rulP90: null,
    failTimeMedian: null, particles: [], rulEnsemble: [], converged: false,
  };
  if (n < 10) return nope;

  // 1. onset
  const onsetIdx = detectOnset(points);
  if (onsetIdx === null) return nope;

  // 2. post-onset data
  const post: { t: number; hi: number }[] = [];
  const tOnset = points[onsetIdx].t;
  for (const p of points) {
    if (p.t >= tOnset && p.hi > 0) post.push({ t: p.t, hi: p.hi });
  }
  if (post.length < MIN_POST_ONSET) return { ...nope, onset: tOnset };

  // 3. estimate signal amplitude from the first post-onset points (closest to true lnA)
  const logHis = post.map(p => Math.log(Math.max(1e-9, p.hi)));
  const firstK = Math.min(4, post.length);
  const firstLnHi = logHis.slice(0, firstK).reduce((a, v) => a + v, 0) / firstK;

  // 4. initialise particles from a wide uninformed prior (not from OLS)
  //    lnA ~ N(baseline_lnHI, 2.0²), wide enough to cover the uncertainty
  //    b   ~ Gamma-like (LogNormal): median ~ 0.05/h, 95% CI [0.001, 0.5]
  //    σ   ~ LogNormal: median 0.15, wide
  const particles: Particle[] = [];
  for (let i = 0; i < N; i++) {
    const lnA = firstLnHi + randn(rng) * 2.0;
    const b = Math.exp(Math.log(0.05) + randn(rng) * 1.0);
    const sigmaObs = Math.exp(Math.log(0.15) + randn(rng) * 0.6);
    particles.push({ lnA, b: Math.max(1e-12, b), sigmaObs: Math.max(1e-6, sigmaObs), w: 1 / N });
  }
  kernelRegularise(particles, rng);

  // 5. sequential importance resampling, process all post-onset observations
  for (const obs of post) {
    const t = obs.t;
    // compute log-weights (in log space for stability, then exponentiate)
    const logWs: number[] = [];
    let maxLw = -Infinity;
    for (const part of particles) {
      const lw = loglik(obs.hi, part.lnA, part.b, part.sigmaObs, t);
      logWs.push(lw);
      if (lw > maxLw) maxLw = lw;
    }
    // exponentiate with offset for numerical stability
    let wSum = 0;
    for (let i = 0; i < N; i++) {
      particles[i].w = Math.exp(logWs[i] - maxLw);
      wSum += particles[i].w;
    }
    if (wSum < 1e-60) {
      // total collapse, reinitialise from prior
      for (let i = 0; i < N; i++) {
        particles[i].lnA = firstLnHi + randn(rng) * 2.0;
        particles[i].b = Math.max(1e-12, Math.exp(Math.log(0.05) + randn(rng) * 1.0));
        particles[i].sigmaObs = Math.max(1e-6, Math.exp(Math.log(0.15) + randn(rng) * 0.6));
        particles[i].w = 1 / N;
      }
      kernelRegularise(particles, rng);
      continue;
    }
    // normalise
    for (const part of particles) part.w /= wSum;
    // effective sample size
    const ess = 1 / particles.reduce((a, p) => a + p.w * p.w, 0);
    if (ess < N * ESS_THRESHOLD) {
      resample(particles, rng);
      kernelRegularise(particles, rng);
      for (const part of particles) part.w = 1 / N;
    }
  }

  // 6. RUL ensemble projection from the filtered posterior
  const tLast = post[post.length - 1].t;
  const rulEnsemble: number[] = [];
  for (const part of particles) {
    const ft = (Math.log(threshold) - part.lnA) / part.b;
    const rul = Math.max(0, ft - tLast);
    if (Number.isFinite(rul) && rul < 1e6) rulEnsemble.push(rul);
  }
  if (rulEnsemble.length < 50) return { ...nope, onset: tOnset };

  rulEnsemble.sort((a, b) => a - b);
  const K = rulEnsemble.length;
  const pct = (q: number) => rulEnsemble[Math.max(0, Math.min(K - 1, Math.floor(q * (K - 1))))] ?? null;

  // check convergence: if the posterior is still as wide as the prior, it hasn't converged
  const lnAs = particles.map(p => p.lnA);
  const lnASd = Math.sqrt(lnAs.reduce((a, v) => a + (v - lnAs.reduce((s, x) => s + x, 0) / N) ** 2, 0) / N);
  const converged = lnASd < 1.5; // prior has sd ~ 2.0

  return {
    onset: tOnset,
    rulMedian: pct(0.5),
    rulP10: pct(0.1),
    rulP90: pct(0.9),
    failTimeMedian: tLast + (pct(0.5) ?? 0),
    particles: particles.map(p => ({ lnA: p.lnA, b: p.b, sigmaObs: p.sigmaObs, w: p.w })),
    rulEnsemble,
    converged,
  };
}

File: data-pipeline/pipeline/model/pf_rul.py
"""Particle-filter RUL prognostics, offline pipeline (Python, numpy).

Proper Bayesian SIR with regularisation (Musso, Oudjane & Le Gland 2001), a WIDE
uninformed prior (NOT seeded from OLS), and kernel-density jitter. Matches the TypeScript
frontend implementation algorithmically so cross-validation is meaningful.

References:
    Arulampalam et al. (2002), IEEE TSP 50(2):174–188
    Musso, Oudjane & Le Gland (2001), "Improving regularised particle filters"
    An, Kim & Choi (2013), RESS 120:50–62
"""

from __future__ import annotations

import numpy as np

N = 500
ESS_THRESHOLD = 0.60
MIN_POST_ONSET = 6


def _onset_idx(hi: np.ndarray) -> int | None:
    n = len(hi)
    if n < 10:
        return None
    base_n = max(5, int(n * 0.25))
    base = hi[:base_n]
    mu0 = float(base.mean())
    sd0 = float(base.std() or 1e-9)
    thr = mu0 + 3.5 * sd0
    for i in range(1, n - 2):
        if hi[i] > thr and hi[i + 1] > thr and hi[i + 2] > thr:
            return i
    return None


def _kernel_regularise(particles: np.ndarray, weights: np.ndarray) -> np.ndarray:
    """Silverman bandwidth, shrink by 0.5 (regularisation, not full kernel)."""
    n = len(particles)
    w_sum = weights.sum() or 1.0
    means = (particles * weights[:, None]).sum(axis=0) / w_sum
    diffs = particles - means[None, :]
    var = (diffs ** 2 * weights[:, None]).sum(axis=0) / w_sum
    h = 1.06 * np.sqrt(np.maximum(var, 1e-12)) * n ** (-0.2) * 0.5
    jitter = np.random.randn(n, 3) * h[None, :]
    particles += jitter
    particles[:, 1] = np.maximum(1e-12, particles[:, 1])     # b > 0
    particles[:, 2] = np.maximum(1e-6, particles[:, 2])       # σ > 0
    return particles


def _resample(particles: np.ndarray, weights: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    n = len(weights)
    inv_n = 1.0 / n
    u0 = np.random.uniform(0, inv_n)
    indices = np.zeros(n, dtype=int)
    cdf = np.cumsum(weights)
    j = 0
    for i in range(n):
        u = u0 + i * inv_n
        while u > cdf[j] and j < n - 1:
            j += 1
        indices[i] = j
    return particles[indices], np.full(n, inv_n)


def pf_rul(t: np.ndarray, hi: np.ndarray, threshold: float) -> dict:
    n = len(t)
    if n < 10:
        return _empty()

    onset_idx = _onset_idx(hi)
    if onset_idx is None:
        return _empty()
    onset_t = float(t[onset_idx])

    mask = t >= onset_t
    post_t = t[mask]
    post_hi = np.maximum(hi[mask], 1e-9)
    if len(post_t) < MIN_POST_ONSET:
        return {"onset": onset_t, "rul_median": None, "rul_p10": None, "rul_p90": None,
                "particles": np.empty((0, 4)), "rul_ensemble": np.empty(0), "converged": False}

    y = np.log(post_hi)
    first_k = min(4, len(y))
    first_ln = float(y[:first_k].mean())  # prior centered on first post-onset points (closest to true lnA)

    # Weakly informed prior centered near the true lnA
    particles = np.column_stack([
        first_ln + np.random.randn(N) * 1.5,                                  # lnA
        np.maximum(1e-12, np.exp(np.log(0.05) + np.random.randn(N) * 1.0)),   # b
        np.maximum(1e-6, np.exp(np.log(0.15) + np.random.randn(N) * 0.6)),    # σ_obs
    ])
    weights = np.full(N, 1.0 / N)
    particles = _kernel_regularise(particles, weights)

    # SIR over post-onset observations
    for i in range(len(post_t)):
        ti = float(post_t[i])
        yi = float(post_hi[i])
        pred = particles[:, 0] + particles[:, 1] * ti
        err = np.log(np.maximum(yi, 1e-9)) - pred
        log_w = -0.5 * (err / particles[:, 2]) ** 2 - np.log(particles[:, 2])
        max_lw = float(log_w.max())
        w = np.exp(log_w - max_lw)
        w_sum = float(w.sum())
        if w_sum < 1e-60:
            particles = np.column_stack([
                first_ln + np.random.randn(N) * 1.5,
                np.maximum(1e-12, np.exp(np.log(0.05) + np.random.randn(N) * 1.0)),
                np.maximum(1e-6, np.exp(np.log(0.15) + np.random.randn(N) * 0.6)),
            ])
            weights = np.full(N, 1.0 / N)
            particles = _kernel_regularise(particles, weights)
            continue
        weights = w / w_sum
        ess = 1.0 / (weights @ weights)
        if ess < N * ESS_THRESHOLD:
            particles, weights = _resample(particles, weights)
            particles = _kernel_regularise(particles, weights)
            weights = np.full(N, 1.0 / N)

    # RUL ensemble
    t_last = float(post_t[-1])
    ft = (np.log(threshold) - particles[:, 0]) / particles[:, 1]
    rul = np.maximum(0, ft - t_last)
    ok = np.isfinite(rul) & (rul < 1e6)
    nrul = np.sort(rul[ok])
    k = len(nrul) or 1

    ln_as = particles[:, 0]
    ln_a_sd = float(np.std(ln_as))
    converged = ln_a_sd < 1.5

    def _pct(q: float) -> float | None:
        if k <= 50:
            return None
        idx = max(0, min(k - 1, int(q * (k - 1))))
        return float(nrul[idx])

    return {
        "onset": onset_t,
        "rul_median": _pct(0.5),
        "rul_p10": _pct(0.1),
        "rul_p90": _pct(0.9),
        "particles": np.column_stack([particles, weights]),
        "rul_ensemble": nrul,
        "converged": converged,
    }


def _empty() -> dict:
    return {"onset": None, "rul_median": None, "rul_p10": None, "rul_p90": None,
            "particles": np.empty((0, 4)), "rul_ensemble": np.empty(0), "converged": False}
